exonum/src/storage/sparse_list_index.rs

```diff
@@ -0,0 +1,838 @@
+/// See its documentation for more.
+///
+/// [`values`]: struct.ListIndex.html#method.values
+/// [`SparseListIndex`]: struct.SparseListIndex.html
+#[derive(Debug)]
+pub struct SparceListIndexValues<'a, V> {
+    base_iter: BaseIndexIter<'a, (), V>,
+}
+
+impl<T, V> SparseListIndex<T, V> {
+    /// Creates a new index representation based on the common prefix of its keys and storage view.
+    ///
+    /// Storage view can be specified as [`&Snapshot`] or [`&mut Fork`]. In the first case only
+    /// immutable methods are available. In the second case both immutable and mutable methods are
+    /// available.
+    /// [`&Snapshot`]: ../trait.Snapshot.html
+    /// [`&mut Fork`]: ../struct.Fork.html
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let snapshot = db.snapshot();
+    /// let prefix = vec![1, 2, 3];
+    /// let index: SparseListIndex<_, u8> = SparseListIndex::new(prefix, &snapshot);
+    /// # drop(index);
+    /// ```
+    pub fn new(prefix: Vec<u8>, view: T) -> Self {
+        SparseListIndex {
+            base: BaseIndex::new(prefix, view),
+            size: Cell::new(None),
+            _v: PhantomData,
+        }
+    }
+}
+
+impl<T, V> SparseListIndex<T, V>
+where
+    T: AsRef<Snapshot>,
+    V: StorageValue,
+{
+    fn size(&self) -> SparseListSize {
+        if let Some(size) = self.size.get() {
+            return size;
+        }
+        let size = self.base.get(&()).unwrap_or_default();
+        self.size.set(Some(size));
+        size
+    }
+
+    /// Returns an element at that position or `None` if out of bounds or it does not exist.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    /// assert_eq!(None, index.get(0));
+    ///
+    /// index.push(42);
+    /// assert_eq!(Some(42), index.get(0));
+    /// index.push(1);
+    /// index.remove(0);
+    /// assert_eq!(None, index.get(0));
+    /// assert_eq!(Some(1), index.get(1));
+    /// ```
+    pub fn get(&self, index: u64) -> Option<V> {
+        self.base.get(&index)
+    }
+
+    /// Returns `true` if the list contains no elements.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    /// assert!(index.is_empty());
+    ///
+    /// index.push(42);
+    /// assert!(!index.is_empty());
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+
+    /// Returns the total amount of elements (including "empty" elements) in the list. In fact
```

Comment:
> In fact capacity reflects total amount of elements that have ever been put to the index and never gets reduced.

It's not entirely true because `set` method with index argument that is outside of capacity can break this statement.

I would say something like:

> The value of capacity is determined by the maximum index of the element ever inserted into the index.